### PR TITLE
🏃 Port E2E test automation from release-0.4 branch

### PR DESCRIPTION
**Pull request by sirao**
Ported E2E test automation added in release-0.4 branch to master branch.
This automation includes conformance tests also.

Verified the test results in local execution and all tests succeeded.

### CODE REVIEW

File: test/e2e/conformance_test.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
+// +build e2e
```
Comment (detiber) [inline test/e2e/conformance_test.go]: This PR will also need the `/scripts/ci-e2e-conformance.sh` file from the release-0.4 branch to fully wire up the conformance test into prow

**Comment by detiber** [comment]
@sirao please address the verify test error:
```
Boilerplate header is wrong for: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/conformance_test.go
make: *** [Makefile:385: verify-boilerplate] Error 1
```

**Comment by sirao** [comment]
> @sirao please address the verify test error:

Addressed the verify test error. pull-cluster-api-provider-aws-verify job succeeded now.

File: scripts/ci-e2e-conformance.sh [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+# shellcheck source=../hack/ensure-go.sh
+source "${REPO_ROOT}/hack/ensure-go.sh"
+# shellcheck source=../hack/ensure-kind.sh
+source "${REPO_ROOT}/hack/ensure-kind.sh"
+# shellcheck source=../hack/ensure-kubectl.sh
+source "${REPO_ROOT}/hack/ensure-kubectl.sh"
+# shellcheck source=../hack/ensure-kustomize.sh
+source "${REPO_ROOT}/hack/ensure-kustomize.sh"
+
```
Comment (detiber) [inline scripts/ci-e2e-conformance.sh]:
```suggestion
# our exit handler (trap)
cleanup() {
  # stop boskos heartbeat
  [[ -z ${HEART_BEAT_PID:-} ]] || kill -9 "${HEART_BEAT_PID}"
}
trap cleanup EXIT

```

To align with #1518

File: scripts/ci-e2e-conformance.sh [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  echo "The provided AWS key is not allowed" 1>&2
+  exit 1
+fi
+
+make test-conformance
+test_status="${?}"
+
+# If Boskos is being used then release the AWS account back to Boskos.
+[ -z "${BOSKOS_HOST:-}" ] || hack/checkin_account.py
+[[ -z ${HEART_BEAT_PID:-} ]] || kill -9 "${HEART_BEAT_PID}"
```
Comment (detiber) [inline scripts/ci-e2e-conformance.sh]:
```suggestion
```

To align with #1518

File: scripts/ci-e2e-conformance.sh (hunk 2 above)
Comment (sirao) [reply scripts/ci-e2e-conformance.sh]: Added the suggested piece of script.

File: scripts/ci-e2e-conformance.sh (hunk 3 above)
Comment (sirao) [reply scripts/ci-e2e-conformance.sh]: Removed the line.

**Comment by detiber** [comment]
It looks like the e2e tests are currently hitting the 4 hour timeout. I will attempt to dig into it a bit further to see if I can see why tomorrow.

**Comment by randomvariable** [comment]
Still hitting the 4h limit.

**Comment by sirao** [comment]
> Still hitting the 4h limit.

Because all the tests are failing. If any one test succeeds, the time will come below 4h. Waiting on fix for https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1530

**Comment by randomvariable** [comment]
Think you need to bump up the Kubernetes version so it picks up the newer AMIs with the aws CLI in them, e.g.

https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1563/files#diff-61994e503eedf336006b7cb84a0fa60aR97

**Comment by detiber** [comment]
I suspect the issue you are now hitting is related to the kustomize changes that were recently made for the v1alpha3 release...  @randomvariable would you be able to help @sirao sort this out?

**Comment by randomvariable** [comment]
@detiber Sure, will take a look this morn (GMT)

**Comment by randomvariable** [comment]
I've updated this branch locally @sirao 

Going to close this one, open a new PR and co-sign the commits.

**Pull request closed without merging**